### PR TITLE
Add hybrid partitioning notebook

### PR DESCRIPTION
**Pull request by SCiarella**
Targets issue #98.

The main thing added here is a notebook that demonstrates that the partitioning component can be replaced by a PyTorch model of the form:

```python
class PartitioningNN(torch.nn.Module):
    def __init__(self, hidden_size=32):
        super().__init__()

        self.trunk = torch.nn.Sequential(
            torch.nn.Linear(1, hidden_size),
            torch.nn.ReLU(),
        )
        ...
     
    def forward(self, dvs):
         ...
```

Instead of introducing a reusable `ml_models/crop/partitioning.py` module, I defined `PartitioningNN` directly in the notebook. I prefer this solution because the NN design is very problem specific, and it does not make sense to have it hard coded in the source. Instead, I would leave it to the user as an external tool. We should maybe provide some template or interface but I am not sure how to do that best. 

For example in this notebook, if I use a simple MLP (ReLU+Linear) it does not learn anything, and instead I had to design a slightly more refined network with two output heads, that closely mimics the behavior of partitioning. 

For the integration into WOFOST, I subclassed `Wofost72` and replaced `self.part` with a small adapter that wraps the neural network as a `SimulationObject`. That means the notebook stays close to the current PCSE/diffWOFOST structure: the NN still publishes the expected partitioning variables to the kiosk and the rest of the crop model can run unchanged.

In other words, the implementation is like:

```python
class HybridWofost72(Wofost72):
    def initialize(...):
        super().initialize(...)
        self.part = PartitionAdapter(..., partition_net=self.partition_net)
```

so notice that the neural network `partition_net` needs an extra wrapper in order to behave like a wofost module and be usable in this way.

Finally, the notebook uses a potential-production dataset and trains the neural partitioning block against the reference `LAI` trajectory, and the specific 'output' of the partitioning (`FR`, `FL`, `FS`, `FO`) are only used as comparison at the end.
It is not fully clear to me if this is the use case or not.

Additionally, I have used as input for `PartitioningNN` only the `DVS` variable, because this is the behavior of the original partitioning model, but nothing prevents us from designing a NN that uses everything that we can find in the `kiosk` instead.

In conclusion, notice that from a numerical point of view it may be more beneficial to calibrate the internal parameters of partitioning as proposed in #73, because a general non-tailored neural network adds a lot of redundant parameters, that at least for this example are not needed and do not improve the results, while slowing down the optimization.  

@SarahAlidoost what do you think?

### CODE REVIEW

**Comment by SCiarella** [comment]
@SarahAlidoost thanks for the great suggestions!
I have changed this branch in the following way:

1. I added the ML partitioning module under `diffwofost.ml_models.crop.partitioning`, including a `PartitioningNN` model and a `DVS_Partitioning_NN` wrapper. This is where the researcher could define their own.
2. I changed the behavior of configuration so crop modules can be customized from `Configuration`. In practice this means `Configuration` now accepts `CROP_KWARGS`, then `Engine` forwards them to the crop constructor, and `Wofost72` can initialize its internal components through a generic `component_overrides` mechanism.
3. I implemented this `component_ovverides` mechanism such that all default WOFOST72 crop components still work as before, but they can now be swapped with alternative implementations, including NNs, through configuration. Let me know in particular what you think about this strategy!
4. Updated the notebook to compare two NN models
5. Fixed broken tests and add a few new ones for the partitioningNN

**Comment by SCiarella** [comment]
@SarahAlidoost 
The notebook has been updated in order to use the same data as [the pcse notebook](https://github.com/ajwdewit/pcse_notebooks/blob/master/11%20Optimizing%20partitioning%20in%20a%20PCSE%20model.ipynb)

File: src/diffwofost/physical_models/engine.py [hunk 1]
```diff
@@ -169,7 +169,25 @@ def _on_CROP_START(
...
+        self.flag_crop_finish = False
+
+        self.crop.finalize(day)
+        self._save_summary_output()
+
+        if self.flag_crop_delete:
+            self.flag_crop_delete = False
+            self.crop._delete()
+            self.crop = None
+            gc.collect()
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/engine.py]:
```suggestion
```

Comment (SarahAlidoost) [reply src/diffwofost/physical_models/engine.py]: It seems that this function is duplicated. This function exist already in engine.

File: src/diffwofost/physical_models/crop/wofost72.py [hunk 2]
```diff
@@ -153,31 +166,67 @@ def initialize(
...
+        replacement component such as a neural partitioning module to consume a
+        trained or trainable PyTorch model while the rest of WOFOST remains
+        unchanged.
         """
         self.params = self.Parameters(parvalues, shape=shape)
         self.rates = self.RateVariables(
             kiosk, publish=["DMI", "ADMI", "REALLOC_LV", "REALLOC_ST", "REALLOC_SO"], shape=shape
         )
         self.kiosk = kiosk
+        component_overrides = self._normalize_component_overrides(component_overrides)
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/wofost72.py]: The function `_normalize_component_overrides` can be moved to utils as `normalize_component_overrides`. This makes the script wofost72 cleaner.

From a design perspective, it would be cleaner if we can call this function in `Engine`. That’ll take a bit of exploration and redesigning the `Engine`. So better to handle it in a separate PR.

File: src/diffwofost/physical_models/crop/wofost72.py [hunk 3]
```diff
@@ -153,31 +166,67 @@ def initialize(
...
-        self.evtra = Evapotranspiration(day, kiosk, parvalues, shape=shape)
-        self.ro_dynamics = Root_Dynamics(day, kiosk, parvalues, shape=shape)
-        self.st_dynamics = Stem_Dynamics(day, kiosk, parvalues, shape=shape)
-        self.so_dynamics = Storage_Organ_Dynamics(day, kiosk, parvalues, shape=shape)
-        self.lv_dynamics = Leaf_Dynamics(day, kiosk, parvalues, shape=shape)
+        for component_name, (attribute_name, _) in self.COMPONENT_SPECS.items():
+            setattr(
+                self,
+                attribute_name,
+                self._initialize_component(
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/wofost72.py]: The function `_initialize_component` can be moved to utils as `initialize_component`. This makes the script wofost72 cleaner.

And also, see [my previous comment](https://github.com/WUR-AI/diffWOFOST/pull/108/changes#r3109193845) about calling this function in Engine.

File: src/diffwofost/physical_models/config.py [hunk 4]
```diff
@@ -144,6 +144,7 @@ class Configuration:
     """Class to store model configuration from a PCSE configuration files."""
 
     CROP: type[SimulationObject]
+    CROP_KWARGS: dict = field(default_factory=dict)
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/config.py]: The `CROP_KWARG` is a key to swap crop modules. Inside the wofost72, they are called `COMPONENT` and in other places they are called `sub-models`. It would be nice to use one clear name for consistency and readability. I suggest replacing with  `CROP_COMPONENT` .

File: src/diffwofost/physical_models/engine.py [hunk 5]
```diff
@@ -169,7 +169,25 @@ def _on_CROP_START(
         self.parameterprovider.set_active_crop(
             crop_name, variety_name, crop_start_type, crop_end_type
         )
-        self.crop = self.mconf.CROP(day, self.kiosk, self.parameterprovider, shape=self._shape)
+        self.crop = self.mconf.CROP(
+            day,
+            self.kiosk,
+            self.parameterprovider,
+            shape=self._shape,
+            **self.mconf.CROP_KWARGS,
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/engine.py]: See my other comment about renaming `CROP_KWARG` to `CROP_COMPONENT`. 

Also I see in the notebook, `CROP_KWARG` is defined as:
```python
hybrid_config.CROP_KWARGS["component_overrides"] = {
    "partitioning": {
        "class": DVS_Partitioning_NN,
        "model": partition_nn,
    }
}
```

Do we need the key `"component_overrides"`, can we remove this from the dictionary as:

```python
hybrid_config.CROP_KWARGS["partitioning"] = {
        "class": DVS_Partitioning_NN,
        "model": partition_nn,
}
```

**Review by SarahAlidoost — CHANGES_REQUESTED** [review]
@SCiarella nice work, well done! 🥇 I like the implementation of `ml_models/crop/partitioning.py`. This is also what I thought: the ml model replaces the physical model inside the "simulation object". I left a few comments. 

My main concern is the design, specifically, doing the swapping in the `Engine` instead of inside `wofost72`. That’ll need some exploration when we revisit the Engine design. I’ll open an issue so we can brainstorm it later.

I was able to run the notebook. Compared to the PCSE one, the results look good. The notebook’s pretty long mostly because of data loading, plotting, and setting up comparisons between three models. That said, I think some codes could move into the Engine or utils, like converting outputs to tensors, but that’s a minor thing

Although the main goal of this PR was just to test swapping a component and build a quick prototype, not to actually train an nn model, it would be nice to get some feedback from the research group on the notebook, something to mention in the next update meeting.

File: src/diffwofost/physical_models/crop/wofost72.py (hunk 2 above)
Comment (SCiarella) [reply src/diffwofost/physical_models/crop/wofost72.py]: Good Idea! for now I will move it to `utils` and later the plan will be to incorporate it into the `Engine`

File: src/diffwofost/physical_models/engine.py (hunk 5 above)
Comment (SCiarella) [reply src/diffwofost/physical_models/engine.py]: done

**Comment by SCiarella** [comment]
@SarahAlidoost thanks for the comments. I have implemented all the changes requested.

I suggest that we merge this as soon as you agree with the changes, so we can mention on slack this new ML-hybrid prototype that can be tested before next meeting, and get some feedback.

**Review by SarahAlidoost — APPROVED** [review]
@SCiarella thanks for addressing the comments! 👍

**Comment by SarahAlidoost** [comment]
> @SarahAlidoost thanks for the comments. I have implemented all the changes requested.
> 
> I suggest that we merge this as soon as you agree with the changes, so we can mention on slack this new ML-hybrid prototype that can be tested before next meeting, and get some feedback.

Good suggestions! it is merged.
